### DISCUSSION TITLE
✅ Refined inputs

### DISCUSSION THREAD

**Discussion by johngeorgewright** [RobinTail/express-zod-api#600] [Ideas]
Are there any future plans to allow refining of an input, and not just it's properties?

For example, I want to check that an input has **at least** one property, but I don't care which one.

```ts
endpointsFactory.build({
  method: "post",
  input: z
    .object({
      email: z.string().email().optional(),
      id: z.string().optional(),
      otherThing: z.string().optional()
    })
    .refine(
      x => Object.keys(x).length >= 1,
      'Please provide at least one property'
    ),
  output: z.object({
    /* ... */
  }),
  handler: async ({ input, logger }) => {
    /* ... */
  },
});
```

Currently the above will fail as the `input` is of type `ZodEffects` which cannot be assigned to `IOSchema`.

Or maybe there's a better solution to the above example

**Comment by RobinTail** [comment]
This is an interesting idea for further development. I will consider the possibility of its implementation in the future.
But for now, you can consider the following approach.
You want to check for at least one property that appears to be a property of something. For example, this is authorization using an email, phone number or token. Surely you can somehow name this entity. In this case: `credentials`.

```ts
endpointsFactory.build({
  method: "post",
  input: z.object({
      credentials: z.object({
        email: z.string().email().optional(),
        phoneNumber: z.string().optional(),
        token: z.string().optional()
      }).refine(
        credentials => Object.keys(credentials).length >= 1,
        'Please provide at least one property'
      ),
    }),
   // ...
});
```

**Comment by ssteuteville** [comment]
I don't think the proposed solution is great tbh, I just opened a similar issue before I saw this. Would be great if there was a way to support this without changing the API input.

https://github.com/RobinTail/express-zod-api/issues/654

**Comment by RobinTail** [comment]
I'm thinking on this feature currently.
But meanwhile I found out that in some cases the desired refinements may turn out to aim achieving the same goal as the Discriminated Unions, that are already supported and allowed on the top level of IO schema. Check 'em out just in case:
https://github.com/colinhacks/zod#discriminated-unions 

CC @ssteuteville

**Comment by RobinTail** [comment]
@johngeorgewright && @ssteuteville ,
the feature released in version 8.3.0 🚀
